### ISSUE TITLE
DTS HCL report does not dump all logs and system configurations

### ISSUE THREAD

**Issue by TomaszAIR** [Dasharo/dasharo-issues#231]
**The problem you're addressing (if any)**

HCL dumps from Alderlake laptops are missing the HDA verbs which are needed to port coreboot. The reason is probably missing drivers, e.g. snd-hda-intel and many others. There may be more undetected missing information which poses a threat of losing data when onboarding new hardware since the dumps are not complete. Dumping procedure based on DTS is not trustworthy right now.

**Describe the solution you'd like**

N/A

**Where is the value to a user, and who might that user be?**

**Describe alternatives you've considered**

**Additional context**

Hint: Using meta-intel provided machine config instead of generic x86 target may save some effort.

**Comment by esesyn01** [comment]
@TomaszAIR I managed to rebuild the DTS image to use machine config `intel-skylake-64` (as it was the only fitting machine) from `meta-intel.` However, the report is very similar to the one from genericx86 config (see the last two reports on the cloud). ROM binary has not been created

**Comment by TomaszAIR** [comment]
@esesyn01 can we compare generated kernel configs with each other? Are there significant changes?

**Comment by esesyn01** [comment]
@TomaszAIR How can I obtain the kernel configs the fastest way?

**Comment by esesyn01** [comment]
@TomaszAIR I compared the two kernel configs, and there are many differences between them: https://www.diffchecker.com/e558RHNV
On the left is the config from generic machine, on the right from the intel machine

**Comment by TomaszAIR** [comment]
e.g. inside workdir of given kernel build, inside image/boot/ there should be config file

**Comment by TomaszAIR** [comment]
@esesyn01 are there drivers enabled like `snd-hda-intel` etc?

**Comment by esesyn01** [comment]
@TomaszAIR `CONFIG_SND_HDA_INTEL=m` so this is as module currently

**Comment by TomaszAIR** [comment]
@esesyn01 so what we should do now is to check how many similar drivers are enabled as modules and maybe rebuild kernel with them compiled in, so set as `y`. If the size would not be too big we could live with that.

The report is now similar probably because the modules were not loaded.

**Comment by kotylamichal** [comment]
I try to reproduce generating logs on Ubuntu, here is a list of software to install on clean version of Ubuntu:

```
i2c-tools
flashrom
ectool
intelp2m
inteltool
msrtool
superiotool
touchpad-info
```

At this time I know that:

- `can't get debug descriptor: Resource temporarily unavailable` in `lsusb.err.log` seems to be ok - I got this same error on Ubuntu as in DTS-CE

**Comment by esesyn01** [comment]
@TomaszAIR I'm now testing this on many setups the results after the step:
```
# echo "Dumping Audio devices configuration..."
for x in /sys/class/sound/card0/hw*; do cat "$x/init_pin_configs" > logs/pin_"$(basename "$x")"; done
for x in /proc/asound/card0/codec#*; do cat "$x" > logs/"$(basename "$x")"; done
```
are:
- On a machine based on `genericx86` without `SND_HDA_INTEL' and `intel-skylake-64` with `SND_HDA_INTEL` and many other sound kernel configs:
```
cat: '/sys/class/sound/card0/hw*/init_pin_configs': No such file or directory
cat: '/proc/asound/card0/codec#*': No such file or directory
```
- On a machine based on `genericx86` with 'SND_HDA_INTEL` and `intel-skylake-64` without `SND_HDA_INTEL`:
```
cat: '/sys/class/sound/card0/hw*/init_pin_configs': No such file or directory
```

I don't know how to go any further. I tried to get information about which driver enables it, but without success.

**Comment by TomaszAIR** [comment]
Basically what we want to do here is to close the gap between report generated with DTS and the one that can be achieve e.g. on Ubuntu (where kernel has more drivers). So maybe let's try do that? Please generate report on these two systems (on Ubuntu you will need couple of programs used in report, you can compile them or use the ones compiled for DTS) and compare them. This could give us a hint which drivers are missing. We should also compare kernel configs

**Comment by macpijan** [comment]
Don't waste time installing tools - there is container containing them: https://github.com/3mdeb/fwdump-docker

**Comment by macpijan** [comment]
But we already had sound devices present, so why having this problem again?

Weren't these added by:

```
CONFIG_SND_HDA_HWDEP=y
CONFIG_SND_HDA_RECONFIG=y
```

?

**Comment by macpijan** [comment]
You know you can grep the kernel sources for the sysfs entries (such as: `init_pin_configs`) you need and find drivers in the source code this way. 
You should also look at help text in Kconfig options: https://elixir.bootlin.com/linux/latest/source/sound/pci/hda/Kconfig#L62

I hope it helps here.

If these are still missing for recent hardware (e.g. present for Tiger Lake, but not for Alder Lake), it most likely means we need kernel upgrade as well.

**Comment by TomaszAIR** [comment]
@macpijan `5.15.36`

**Comment by macpijan** [comment]
It might be OK. At least the hda_intel.c already has PCI IDs for Alder Lake.

The only Alder Lake configs I can see that might be worth checking are:
- PINCTRL_ALDERLAKE
- SND_SOC_SOF_ALDERLAKE

**Comment by esesyn01** [comment]
Finally, the missing files for the soundcard are present after adding configs for Alder lake.
However, the ROM binary file is still not created.

**Comment by esesyn01** [comment]
@macpijan @TomaszAIR From the generated report, there are several errors:
- `flashrom-read.err.log`:
```
WARNING: No chipset found. Flash detection will most likely fail.
Laptops, notebooks and netbooks are difficult to support and we
recommend to use the vendor flashing utility. The embedded controller
(EC) in these machines often interacts badly with flashing.
See the manpage and https://flashrom.org/Laptops for details.

If flash is shared with the EC, erase is guaranteed to brick your laptop
and write may brick your laptop.
Read and probe may irritate your EC and cause fan failure, backlight
failure and sudden poweroff.
You have been warned.
========================================================================
```
- `inteltool.err.log`: `Error: Dumping AMBs on this MCH is not (yet) supported.`
- `i2cdetect.err.log`:
```
Warning: Can't use SMBus Quick Write command, will skip some addresses
Warning: Can't use SMBus Quick Write command, will skip some addresses
```
- `msrtool.err.log`:
```
msrtool 4.16-526-g3ab6316300
Detected system linux: Linux with /dev/cpu/*/msr

Unable to detect a known target; can not decode any MSRs! (Use -t to force)
Please send a report or patch to coreboot@coreboot.org. Thanks for your help!
```
- `lsusb.err.log`:
```
can't get debug descriptor: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
```
For me, the key is the flashrom error. Do you have any thoughts about it?
The whole report: 
[Notebook_NV4xPZ_Dasharo_(coreboot+UEFI)__2022_10_27_12_32_42_219411699.tar.gz](https://github.com/Dasharo/dasharo-issues/files/9879559/Notebook_NV4xPZ_Dasharo_.coreboot%2BUEFI.__2022_10_27_12_32_42_219411699.tar.gz)

**Comment by macpijan** [comment]
Just use most recent flashrom revision https://github.com/Dasharo/flashrom

**Comment by esesyn01** [comment]
@macpijan @TomaszAIR @kotylamichal Here is the report made on ubuntu on the same device:
[Notebook_NV4xPZ_Dasharo_(coreboot+UEFI)_.tar.gz](https://github.com/Dasharo/dasharo-issues/files/9879863/Notebook_NV4xPZ_Dasharo_.coreboot%2BUEFI._.tar.gz)

Changes between Ubuntu and DTS:
- On Ubuntu, there is a present log file for Intel Alderlake-P HDMI codec with corresponding pinout file. There are no such files in report from DTS
- There are some differences in `cpuinfo.log` file: https://www.diffchecker.com/VdaBWaos (DTS is on the left)
- On DTS is present `decode-dimms.log` file
- Dmesg files differs
- On DTS file `i2cdetect.log` with corresponding error log is present
- `input-bustypes` log file has more entries on Ubuntu
- There is an error in lscpi: `lspci: Unable to load libkmod resources: error -12` on Ubuntu
- On Ubuntu, there is no information about touchpad, which is present in DTS

Both of them shares errors with msrtool, lsusb, flashrom and inteltool.

To sum up, we should also enable HDMI codecs on Alderlake in DTS

**Comment by esesyn01** [comment]
@TomaszAIR Even if the problem with sound should be gone, it isn't gone entirely. Directly after booting, these files:
```
cat: '/sys/class/sound/card0/hw*/init_pin_configs': No such file or directory
cat: '/proc/asound/card0/codec#*': No such file or directory
```
do not exist. But this problem occurs in a short time after booting. After time passes (about a minute), These files are present, and everything works fine.

Also, I added the missing Intel drivers, like `INTEL_BT`, some additional sound drivers, thermal drivers, etc.

**Comment by kotylamichal** [comment]
@esesyn01 so maybe we could wait until uptime will be higher than 1minute?

with ` echo $(awk '{print $1}' /proc/uptime) / 60 | bc`, rather we do not have added `bc` in meta-dts-ce base image or similar method

**Comment by macpijan** [comment]
This sounds way to hacky.

How about here https://github.com/Dasharo/meta-dts-ce/blob/main/meta-dts-distro/recipes-dts/reports/dasharo-hcl-report/dasharo-hcl-report#L70 before dumping contents of these files, you first check if they exist. You check that in a loop until they exist, with a maximum timeout of 60 seconds.

It is quite likely, that by the time users starts HCL and the previous command in the HCL script finish, these files would already be there, so no extra waiting time (or very short one) is required.

Make sure to leave a comment and  also an issue to track this problem in the future.

**Comment by esesyn01** [comment]
@TomaszAIR I created a PR with a check from @macpijan 
Now I will create the issue and I'm waiting for review
